src/MicrosoftGraphBinding/Bindings/ExcelAsyncCollector.cs

```
@@ -45,11 +46,39 @@ public Task AddAsync(string item, CancellationToken cancellationToken = default(
             {
                 throw new ArgumentNullException("No row item");
             }
-
-            this._rows.Add(JObject.Parse(item));
+            JToken parsedToken = JToken.Parse(item);
+            if (parsedToken is JObject)
+            {
+                _rows.Add(parsedToken as JObject);
+            }
+            else
+            {
+                //JavaScript Array, so add metadata manually
+                var array = parsedToken as JArray;
+                bool arrayIsNested = array.All(element => element is JArray);
+                if (arrayIsNested)
+                {
+                    var consolidatedRow = new JObject();
```

Is the `consolidatedRow` a Graph API thing? Is there a link to documentation for this that we can include as a comment?

---

For the most of the Excel PUT/POST Graph APIs, you can feed it a nested `JArray`, where each internal `JArray` is a row. Currently if the user calls `AddAsync()` repeatedly (like with an `ICollector<Poco>`), each row will be it's own nested `JArray` of size 1, where the internal array is that row, and each one will be fed into it's own call to the Graph API. In the case of `Update`, where the whole Excel spreadsheet is updated, this doesn't work, so we have to consolidate all the rows into one nested `JArray` so that each row doesn't overwrite the last. You could argue that this would be a good idea in the Append case, but I didn't want to touch this more than I had to considering the time frame.

Since the Node case is the same as `out string binding` in C#, the entire nested JArray is already all in one Row, hence why I called it a consolidatedRow. It's all a bit confusing, and at some point I would really like to clean this all up when there is more time.

---

So this might be answering my previous question about what we intend to support and when. I think that's fine, but it would help to add some comments to make it clear what we're intending to support and what we're not. I wonder too if we should throw a friendly exception if someone tries to pass an array of JObjects, which clearly isn't supported yet (e.g. throw if arrayIsNested == false, or something like that).

---
In reply to: [178393317](https://github.com/Azure/azure-functions-microsoftgraph-extension/pull/47#discussion_r178393317) [](ancestors = 178393317)

---

That sounds like a good approach for now. I'll do that quick.